Framework/Muon/src/PSIBackgroundSubtraction.cpp

```diff
@@ -0,0 +1,129 @@
+// SPDX - License - Identifier: GPL - 3.0 +
+#include "MantidMuon/PSIBackgroundSubtraction.h"
+#include "MantidAPI/FunctionFactory.h"
+#include "MantidAPI/IFunction.h"
+#include "MantidAPI/MatrixWorkspace.h"
+#include "MantidKernel/BoundedValidator.h"
+
+using namespace Mantid::API;
+using namespace Mantid::Kernel;
+using namespace Mantid::DataObjects;
+
+namespace Mantid {
+namespace Muon {
+
+namespace {
+constexpr char *MINIMISER = "Levenberg-Marquardt";
+constexpr double FIT_TOLERANCE = 10;
+} // namespace
+
+// Register the algorithm into the AlgorithmFactory
+DECLARE_ALGORITHM(PSIBackgroundSubtraction)
+
+void PSIBackgroundSubtraction::init() {
+
+  declareProperty(
+      std::make_unique<WorkspaceProperty<>>(
+          "InputWorkspace", "", Direction::Input, PropertyMode::Mandatory),
+      "Input workspace containing the PSI bin data "
+      "which the background correction will be applied to.");
+
+  auto mustBePositive = std::make_shared<Kernel::BoundedValidator<int>>();
+  mustBePositive->setLower(0);
+  declareProperty(
+      "MaxIterations", 500, mustBePositive,
+      "Stop after this number of iterations if a good fit is not found");
+}
+
+std::map<std::string, std::string> PSIBackgroundSubtraction::validateInputs() {
+  std::map<std::string, std::string> errors;
+
+  MatrixWorkspace_sptr inputWS = getProperty("InputWorkspace");
+
+  if (inputWS->YUnit() != "Counts") {
+    errors["InputWorkspace"] = "Input Workspace should be a counts workspace.";
+  }
+  return errors;
+}
+
+void PSIBackgroundSubtraction::exec() {
+  setRethrows(true);
+  MatrixWorkspace_sptr inputWS = getProperty("InputWorkspace");
+  // Caclulate and subtract background from inputWS
+  calculateBackgroundUsingFit(*inputWS);
+}
+/**
+ * Calculate the background of a PSI workspace by performing a fit, comprising
+ of a FlatBackground and ExpDecayMuon, on the second half of the PSI data.
+ * @param inputWorkspace ::  Input PSI workspace which is modified inplace
+ */
+void PSIBackgroundSubtraction::calculateBackgroundUsingFit(
+    MatrixWorkspace &inputWorkspace) {
+  IAlgorithm_sptr fit = setupFitAlgorithm(inputWorkspace.getName());
+  auto numberOfHistograms = inputWorkspace.getNumberHistograms();
+  for (size_t index = 0; index < numberOfHistograms; ++index) {
+    auto &countsdata = inputWorkspace.mutableY(index);
+    auto &timedata = inputWorkspace.x(index);
+    auto maxtime = timedata.back();
+    auto [background, fitQuality] =
+        calculateBackgroundFromFit(fit, maxtime, static_cast<int>(index));
+    // If fit quality is poor, do not subtract the background and instead log a
+    // warning
+    if (fitQuality > FIT_TOLERANCE) {
+      g_log.warning()
+          << "Fit quality obtained in PSIBackgroundSubtraction is poor. "
+          << "Skipping background calculation for WorkspaceIndex: "
+          << static_cast<int>(index) << "\n";
+    } else {
+      countsdata = countsdata - background;
```

Comment:
If we change this to a `DataProcessorAlgorithm` then we would need to use `self.createChildAlgorithm('Minus');` here rather than the operator overloads or unfortunately the historyy won't get recorded properly.

Reply:
Can I execute the Minus algorithm with a HistogramData object and a double value, struggling to see how I would make this Minus algorithm work in place of the currently used operator overload. Would I need to create a temporary workspace with the backgrounds for each spectra?

Reply:
Ah yes thanks for bring that up as I originally had missed that this goes spectra-by-spectra so yes a temporary workspace would need to be created - this could be done before hand by creating a single column with each row containing the correct value and then the minus done outside the loop. This will hurt the performance though so then its worth considering if it is worth it?

Reply:
The size of the PSI workspaces is always going to be small (<10 spectra), so I think the performance hit is small. I've had a go at converting it to a DataProcessing algorithm and the workspace history now shows exactly what happened to the workspace, which is definitely a good thing for transparency.